### ISSUE TITLE
TypeError reconstructing datetime on Python 3.6

### ISSUE THREAD

**Issue by jaraco** [jsonpickle/jsonpickle#153]
Starting with Python 3.6, jsonpickle fails to reconstruct a datetime:

```
$ python -m rwt jsonpickle
Collecting jsonpickle
Installing collected packages: jsonpickle
Successfully installed jsonpickle-0.9.3
Python 3.6.0b3 (v3.6.0b3:8345e066c0ed, Oct 31 2016, 18:05:23) 
[GCC 4.2.1 (Apple Inc. build 5666) (dot 3)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import jsonpickle
>>> import datetime
>>> jsonpickle.loads(jsonpickle.dumps(datetime.datetime.now()))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/var/folders/c6/v7hnmq453xb6p2dbz1gqc6rr0000gn/T/rwt-rw7fw0wi/jsonpickle/__init__.py", line 152, in decode
    return unpickler.decode(string, backend=backend, keys=keys)
  File "/var/folders/c6/v7hnmq453xb6p2dbz1gqc6rr0000gn/T/rwt-rw7fw0wi/jsonpickle/unpickler.py", line 26, in decode
    return context.restore(backend.decode(string), reset=reset)
  File "/var/folders/c6/v7hnmq453xb6p2dbz1gqc6rr0000gn/T/rwt-rw7fw0wi/jsonpickle/unpickler.py", line 119, in restore
    value = self._restore(obj)
  File "/var/folders/c6/v7hnmq453xb6p2dbz1gqc6rr0000gn/T/rwt-rw7fw0wi/jsonpickle/unpickler.py", line 159, in _restore
    return restore(obj)
  File "/var/folders/c6/v7hnmq453xb6p2dbz1gqc6rr0000gn/T/rwt-rw7fw0wi/jsonpickle/unpickler.py", line 236, in _restore_object
    instance = handler(self).restore(obj)
  File "/var/folders/c6/v7hnmq453xb6p2dbz1gqc6rr0000gn/T/rwt-rw7fw0wi/jsonpickle/handlers.py", line 173, in restore
    return cls.__new__(cls, *params)
TypeError: an integer is required (got type bytes)
```

I suspect this might be a regression in Python itself.

**Comment by jaraco** [comment]
Indeed, it seems to be violating the reduction protocol.

```
>>> reduction = datetime.datetime.now().__reduce__()
>>> reduction[0](*reduction[1])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: an integer is required (got type bytes)
```

**Comment by jaraco** [comment]
[Reported upstream](http://bugs.python.org/issue28752).

**Comment by davvid** [comment]
Nice, thanks @jaraco, it's not often that our library gets to expose a bug in CPython itself.

Would you recommend that we either have compat code to paper over 3.6, or should we just recommend users not use the library on this version?   From the upstream bug, it sounds like we can workaround it by calling `__reduce_ex__` on 3.6 only.  worth it?

**Comment by jaraco** [comment]
I get the feeling that the issue will be patched in Python 3.6 before it's released, so jsonpickle might not need to do anything. I'll follow the upstream ticket and evaluate their recommendation and make a recommendation here if they don't simply restore `__reduce__` support.

**Comment by jaraco** [comment]
Fixed in Python 3.6b4 just in the nick of time.

```
$ python -V
Python 3.6.0b4
$ python -m rwt jsonpickle
Collecting jsonpickle
Installing collected packages: jsonpickle
Successfully installed jsonpickle-0.9.3
Python 3.6.0b4 (v3.6.0b4:18496abdb3d5, Nov 21 2016, 20:44:47) 
[GCC 4.2.1 (Apple Inc. build 5666) (dot 3)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import jsonpickle
>>> import datetime
>>> jsonpickle.loads(jsonpickle.dumps(datetime.datetime.now()))
datetime.datetime(2016, 11, 22, 16, 51, 54, 667612)
```
